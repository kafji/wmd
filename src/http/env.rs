use crate::{
    app::SearchTarget,
    search_targets::SearchTargets,
    url_maker::UrlMaker,
    templating::{create_templates, Templates},
};
use anyhow::Error;
use std::sync::Arc;
use url::Url;

/// HTTP server environment.
#[derive(Clone)]
pub struct Env(Arc<Inner>);

struct Inner {
    base_url: Url,
    templates: Templates,
    url_maker: UrlMaker,
    search_targets: Vec<SearchTarget>,
    search_targets2: SearchTargets,
}

impl Env {
    pub fn new(base_url: Url, search_targets: Vec<SearchTarget>) -> Result<Self, Error> {
        let templates = create_templates()?;
        let url_maker = UrlMaker::new(&search_targets)?;
        let search_targets2 = SearchTargets::new(search_targets.clone())?;
        let i = Inner {
            base_url,
            templates,
            url_maker,
            search_targets,
            search_targets2,
        };
        let s = Self(Arc::new(i));
        Ok(s)
    }

    pub fn base_url(&self) -> &Url {
        &self.0.base_url
    }

    pub fn templates(&self) -> &Templates {
        &self.0.templates
    }

    pub fn target_url_maker(&self) -> &UrlMaker {
        &self.0.url_maker
    }

    pub fn search_targets(&self) -> &[SearchTarget] {
        &self.0.search_targets
    }

    pub fn search_targets2(&self) -> &SearchTargets {
        &self.0.search_targets2
    }
}
