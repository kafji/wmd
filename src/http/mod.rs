mod env;
mod handlers;
mod server;

pub use self::env::Env as ServerEnv;
pub use self::server::start as start_server;
