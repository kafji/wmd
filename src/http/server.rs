use super::{env::Env, handlers::*};
use anyhow::Error;
use axum::{routing::get, Extension, Router, Server};
use tower_http::trace::TraceLayer;

pub fn router(env: Env) -> Router {
    Router::new()
        .route("/", get(get_index))
        .route("/search", get(get_search))
        .route("/prefixes", get(get_prefixes))
        .route("/opensearch.xml", get(get_opensearch_xml))
        .route("/robots.txt", get(get_robots_txt))
        .layer(Extension(env))
}

/// Starts server.
pub async fn start(port: u16, env: Env) -> Result<(), Error> {
    let app = router(env).layer(TraceLayer::new_for_http());
    let addr = ([0, 0, 0, 0], port).into();
    Server::bind(&addr).serve(app.into_make_service()).await?;
    Ok(())
}
