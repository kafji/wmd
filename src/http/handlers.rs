/*!
HTTP requests handlers.
*/

use super::env::Env;
use crate::{command_processing, templating::Template::*};
use axum::{
    body::Full,
    extract::Query,
    http::StatusCode,
    response::{Html, IntoResponse, Redirect, Response},
    Extension,
};
use handlebars::RenderError;
use serde::Deserialize;
use serde_json::json;
use std::fmt::Debug;

#[cfg(debug_assertions)]
use axum_macros::debug_handler;

/// Produces UTF-8 HTML reply.
macro_rules! reply_with_html {
    ($html:expr) => {
        return Ok(Html($html).into_response())
    };
    ($tmplts:expr, $tmplt:ident, $tmplt_ctx:expr) => {{
        let html = $tmplts.render($tmplt.name(), $tmplt_ctx)?;
        reply_with_html!(html)
    }};
}

/// Produces UTF-8 plain text reply.
macro_rules! reply_with_text {
    ($text:expr) => {
        return {
            let resp = Response::builder()
                .header("content-type", mime::TEXT_PLAIN_UTF_8.to_string())
                .body(Full::from($text))
                .unwrap();
            Ok(resp.into_response())
        }
    };
    ($tmplts:expr, $tmplt:ident, $tmplt_ctx:expr) => {{
        let text = $tmplts.render($tmplt.name(), $tmplt_ctx)?;
        reply_with_text!(text)
    }};
}

macro_rules! respond_failure {
    ($desc:expr, $code:expr, $display:expr) => {{
        tracing::error!("{}", $desc);
        ($code, $display).into_response()
    }};
}

/// Result of response handler.
type Reply = Result<Response, Error>;

/// Denotes failure when handling response.
#[derive(Debug)]
pub enum Error {
    Renderer { desc: String },
    InvalidTargetUrl { desc: String },
}

impl From<RenderError> for Error {
    fn from(s: RenderError) -> Self {
        Self::Renderer { desc: s.desc }
    }
}

impl IntoResponse for Error {
    fn into_response(self) -> Response {
        match self {
            Error::Renderer { desc } => respond_failure!(
                desc,
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to generate HTML."
            ),
            Error::InvalidTargetUrl { desc } => respond_failure!(
                desc,
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to make target URL."
            ),
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct IndexParams {
    k: Option<String>,
}

#[cfg_attr(debug_assertions, axum_macros::debug_handler)]
pub async fn get_index(Extension(env): Extension<Env>, Query(params): Query<IndexParams>) -> Reply {
    let query: Option<String> = params.k.map(|keywords| {
        let prefix = &env.search_targets2().default().prefix;
        let mut query = String::with_capacity(prefix.len() + 1 + keywords.len());
        query.push_str(prefix);
        query.push(' ');
        query.push_str(&keywords);
        query
    });
    reply_with_html!(
        env.templates(),
        HomePageHtml,
        &json!({
            "query": query,
        })
    )
}

#[derive(Debug, Deserialize)]
pub struct SearchParams {
    q: String,
}

#[cfg_attr(debug_assertions, debug_handler)]
pub async fn get_search(
    Extension(env): Extension<Env>,
    Query(params): Query<SearchParams>,
) -> Reply {
    let registry = env.command_registry();
    let output = command_processing::process(&registry, &params.q);
    let response = match output {
        command_processing::Output::Redirect(url) => {
            let url = url.map_err(|x| Error::InvalidTargetUrl {
                desc: x.to_string(),
            })?;
            Redirect::temporary(url.as_str()).into_response()
        }
        command_processing::Output::Unhandled => {
            // processor can't handle this command, redirect to homepage/search-form-page
            let mut url = env.base_url().clone();
            let params = Some(format!("k={}", &params.q));
            url.set_query(params.as_deref());
            Redirect::temporary(url.as_str()).into_response()
        }
    };
    Ok(response)
}

#[cfg_attr(debug_assertions, debug_handler)]
pub async fn get_prefixes(Extension(env): Extension<Env>) -> Reply {
    let targets = env.search_targets();
    reply_with_html!(
        env.templates(),
        PrefixesPageHtml,
        &json!({
            "prefixes": targets,
        })
    )
}

#[cfg_attr(debug_assertions, debug_handler)]
pub async fn get_opensearch_xml(Extension(env): Extension<Env>) -> Reply {
    let mime_type = r"application/opensearchdescription+xml; charset=utf-8";
    let body = env.templates().render(
        OpensearchXml.name(),
        &json!({
            "baseUrl": env.base_url(),
        }),
    )?;
    Ok(Response::builder()
        .header("content-type", mime_type)
        .body(Full::from(body))
        .unwrap()
        .into_response())
}

#[cfg_attr(debug_assertions, debug_handler)]
pub async fn get_robots_txt(Extension(env): Extension<Env>) -> Reply {
    reply_with_text!(env.templates(), RobotsTxt, &())
}
