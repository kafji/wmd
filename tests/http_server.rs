use http::StatusCode;
use reqwest::redirect::Policy;
use std::time::Duration;
use tokio::{sync::oneshot, task};

macro_rules! launch_server {
    ($ready_rx:ident) => {
        let (ready_tx, $ready_rx) = oneshot::channel();
        let _server = task::spawn(async move {
            let port = getport::reserve_tcp_port();
            let port = port.take();
            let app = tokio::spawn(async move {
                let cfgsrc = wmd::ConfigSource::File("./wmd.example.toml");
                wmd::start_app(port, &[cfgsrc]).await.unwrap();
            });
            tokio::time::sleep(Duration::from_millis(100)).await;
            ready_tx.send(port).unwrap();
            app.await
        });
    };
}

macro_rules! request_once {
    ($path:expr) => {{
        launch_server!(ready_rx);
        let port = ready_rx.await.unwrap();
        let client = reqwest::Client::builder()
            .redirect(Policy::none())
            .build()
            .expect("failed to create HTTP client");
        let response = client
            .get(format!("http://127.0.0.1:{}{}", port, $path))
            .send()
            .await
            .unwrap();
        response
    }};
}

macro_rules! assert_string_response {
    ($res:expr, $status:expr, $mime:expr, $assert_body:expr) => {
        let status = $res.status();
        assert_ne!(status, StatusCode::NOT_FOUND);
        let resp_mime: mime::Mime = {
            let response_mime = $res
                .headers()
                .get("content-type")
                .expect("missing header `content-type`")
                .to_str()
                .unwrap();
            response_mime.parse().unwrap()
        };
        let body = $res.text().await.unwrap();
        $assert_body(body);
        assert_eq!(resp_mime, $mime);
        assert_eq!(status, StatusCode::OK);
    };
}

#[tokio::test]
async fn test_get_index() {
    let response = request_once!("/");

    assert_string_response!(response, StatusCode::OK, mime::TEXT_HTML_UTF_8, |body| {
        insta::assert_display_snapshot!(body);
    });
}

#[tokio::test]
async fn test_get_index_with_param() {
    let response = request_once!("/?k=hello");

    assert_string_response!(response, StatusCode::OK, mime::TEXT_HTML_UTF_8, |body| {
        insta::assert_display_snapshot!(body);
    });
}

#[tokio::test]
async fn test_do_search() {
    {
        // trivial case

        let response = request_once!("/search?q=drx%20tokio");

        assert_eq!(response.status(), StatusCode::TEMPORARY_REDIRECT);
        assert_eq!(
            response.headers().get("location").unwrap(),
            "https://docs.rs/tokio"
        );
    };

    {
        // keywords should be trimmed
        let response = request_once!("/search?q=drx%20tokio%20");

        assert_eq!(response.status(), StatusCode::TEMPORARY_REDIRECT);
        assert_eq!(
            response.headers().get("location").unwrap(),
            "https://docs.rs/tokio"
        );
    };
}

#[tokio::test]
async fn test_do_search_with_missing_prefix() {
    let response = request_once!("/search?q=tokio");

    assert_eq!(response.status(), StatusCode::TEMPORARY_REDIRECT);
    assert_eq!(
        response.headers().get("location").unwrap(),
        "http://127.0.0.1:3000/?k=tokio"
    );
}

#[tokio::test]
async fn test_get_prefixes() {
    let response = request_once!("/prefixes");

    assert_string_response!(response, StatusCode::OK, mime::TEXT_HTML_UTF_8, |body| {
        insta::assert_display_snapshot!(body);
    });
}

#[tokio::test]
async fn test_get_opensearch_xml() {
    let response = request_once!("/opensearch.xml");

    let mime: mime::Mime = r"application/opensearchdescription+xml; charset=utf-8"
        .parse()
        .unwrap();
    assert_string_response!(response, StatusCode::OK, mime, |body| {
        insta::assert_display_snapshot!(body);
    });
}

#[tokio::test]
async fn test_get_robots_txt() {
    let response = request_once!("/robots.txt");

    assert_string_response!(response, StatusCode::OK, mime::TEXT_PLAIN_UTF_8, |body| {
        insta::assert_display_snapshot!(body);
    });
}

// todo(kfj): test ripm
