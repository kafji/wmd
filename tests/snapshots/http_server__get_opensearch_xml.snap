---
source: tests/http_server.rs
assertion_line: 117
expression: body
---
<?xml version="1.0" encoding="utf-8"?>
<OpenSearchDescription xmlns="http://a9.com/-/spec/opensearch/1.1/">
  <ShortName>Wmd</ShortName>
  <Description>Web search proxy.</Description>
  <Url type="text/html" template="http://127.0.0.1:3000//search?q={searchTerms}" />
</OpenSearchDescription>

